connect-socket.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TH connect 1
.SH NAME
connect-socket \- connect a socket and run a program
.SH SYNOPSIS
.B connect-socket
.RI [ options ]
.RB [ -- ]
.I fd
.RI [ family ]
.RI [ type ]
.I address
.RB [ -- ]
.IR command ...
.SH DESCRIPTION
\fBconnect-socket\fR connects a socket and executes a command.
.PP
The arguments are as follows:
.TP 8
.I fd
The file descriptor to use.
.TP
.I family
The socket family.  This may be \fBinet\fR for internet sockets or
\fBunix\fR for UNIX socket.  If the family is omitted, \fBinet\fR is
assumed.
.TP
.I type
This may be \fBstream\fR for stream sockets or \fBdgram\fR for
datagram sockets.  For internet sockets, this maps onto TCP and UDP.
The socket type is optional and defaults to \fBstream\fR.
.TP
.I address
The next argument is the socket address.  The syntax of this depends
on the socket family.
.IP
For internet sockets, the address should either be
\fIaddress\fB:\fIport\fR or just \fIport\fR.  \fIaddress\fR may be
either a hostname or an IP address.  \fIport\fR may be a service name
or a port number.
.IP
For UNIX sockets, the address is the path of the socket.
.TP
.I command
The command to execute.
.SH OPTIONS
.TP
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



connect-socket.c
/* 
   connect-socket - connect a socket and run a command

   Copyright (C) 2001 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/un.h>
#include <unistd.h>
#include <getopt.h>

#include "utils.h"

/* Option flags and variables */
static struct option const long_options[] =
{
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fputs(
"Usage:\n"
"  connect-socket [options] [--] fd [family] [type] address [--] command ...\n"
"\n"
"Options:\n"
"  -h, --help                    Usage message\n"
"  -V, --version                 Version number\n"
"\n"
"Families are \"inet\" or \"unix\".  Types are \"stream\" or \"dgram\".  For\n"
"internet sockets, addresses are \"addr:port\" or just \"port\".  For UNIX\n"
"sockets, the address is the path name.\n"
, fp) < 0)
    fatale("output error");
  exit(exit_status);
}

int main(int argc, char **argv) {
  int n;
  int fdno;
  int fd;
  union {
    struct sockaddr sa;
    struct sockaddr_in in;
    struct sockaddr_un un;
  } u;
  int len, pf, type, proto;
  
  setprogname(argv[0]);
  
  while((n = getopt_long(argc, argv, 
			 "hV",
			 long_options, (int *)0)) >= 0) {
    switch(n) {
    case 'V':
      printf("connect-socket %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);
      
    default:
      usage(stderr, 1);
    }
  }

  if(optind >= argc
     || !isdigit((unsigned char)argv[optind][0]))
    fatal("no file descriptor specified");
    
  fdno = atoi(argv[optind++]);

  /* parse the socket name */
  len = sizeof u;
  parse_socket_arg(&optind, argc, argv, &u.sa,
		   &len, &pf, &type, &proto);

  /* create the socket */
  if((fd = socket(pf, type, proto)) < 0)
    fatale("error creating socket");
  
  
  if(connect(fd, &u.sa, len) < 0)
    fatale("error connecting socket");
  
  /* use the chosen FD */
  if(fd != fdno) {
    dup2_e(fd, fdno);
    close_e(fd);
  }

  /* skip separator */
  if(optind < argc && !strcmp(argv[optind], "--"))
    ++optind;
  
  /* check there's a command */
  if(optind >= argc)
    fatal("no command specified");
  
  /* execute the program */
  if(execvp(argv[optind], argv + optind) < 0)
    fatale("error executing %s", argv[optind]);
  fatal("execvp %s unexpectedly returned", argv[optind]);
}


/*
Local Variables:
c-basic-offset:2
comment-column:40
End:
*/
